lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/HeartBeatManager.cs

```diff
@@ -166,17 +168,30 @@ public void Send(EvaluatorHeartbeatProto evaluatorHeartbeatProto)
                     _queuedHeartbeats.Enqueue(evaluatorHeartbeatProto);
                     LOGGER.Log(Level.Error, string.Format(CultureInfo.InvariantCulture, "Sending heartbeat to driver experienced #{0} failure. Hearbeat queued as: [{1}]. ", _heartbeatFailures, evaluatorHeartbeatProto), e);
 
-                    if (_heartbeatFailures >= _maxHeartbeatRetries)
+                    if (_driverConnection.Get() is MissingDriverConnection)
                     {
-                        LOGGER.Log(Level.Warning, "Heartbeat communications to driver reached max of {0} failures. Driver is considered dead/unreachable", _heartbeatFailures);
-                        LOGGER.Log(Level.Info, "=========== Entering RECOVERY mode. ===========");
-                        ContextManager.HandleDriverConnectionMessage(new DriverConnectionMessageImpl(DriverConnectionState.Disconnected));
-
-                        LOGGER.Log(Level.Info, "instantiate driver reconnect implementation: " + _driverConnection);
-                        _evaluatorSettings.OperationState = EvaluatorOperationState.RECOVERY;
-
-                        // clean heartbeat failure
-                        _heartbeatFailures = 0;
+                        if (_heartbeatFailures >= _maxHeartbeatRetries * DriverConnRetryFactor)
+                        {
+                            Environment.Exit(1);
```

Comment:
I'd be careful with re-throwing exception if `DriverConnection` is implemented. The current code which tries to retrieve driver information in YARN cluster can throw exception if, for example, RM can't be reached - but this can be transient and go away at the next retry. In fact, if evaluator can't talk to a healthy driver chances are high that it won't be able to reach RM at the same time.

I'd throw exception in `OnNext(Alarm)` when `driverConnection.Get()` is not implemented (and throws `NotImplementedException`, which we've observed), and not modify behavior in other cases. This code path is never used (because we don't use high availability in production yet), and breaking changes can go unnoticed for a long time.